### PR TITLE
Add DevSecOps documentation and guidance

### PR DESCRIPTION
**Pull request by LizaShak**
Changes introduced in this PR:

1. Introducing documentation on DevOps Security (DevSecOps) best practices, information and tools for application.
2. Updating the existing Credential Scanning documentation to follow the same template.
3. Linking to DevSecOps section in other relevant documentations.

### CODE REVIEW

**Review by TessFerrandez — APPROVED** [review]
Great add to the playbook, very useful content

File: continuous-integration/dev-sec-ops/secret-management/secrets_rotation.md [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+Secret rotation is the process of refreshing the secrets that are used by the application.
+For further reading see [CSEDevSecOps Secret Management Scenarios](https://github.com/microsoft/CSEDevSecOps/tree/master/Scenarios/SecretManagement)
+
+## Why Secrets Rotation [The Why]
+
+Secrets are an asset and as such have a potential to be leaked or stolen. By rotating the secrets, we are revoking any secrets that may have been compromised. Therefore, secrets should be rotated frequently.
+
+## Applying Secrets Rotation [the how]
+
+To promote frequent rotation of a secret - define an automated periodic secret rotation process.
```
Comment (seushermsft) [inline continuous-integration/dev-sec-ops/secret-management/secrets_rotation.md:12]: It feels like we should have some reference to Managed Identities in this section that mentions how using a Managed Identity allows you to offload secret rotation to Azure, and should be our default choice as long as meets the security/compliance requirements of the customer.

I didn't see Managed Identities called out in the Secret Management Scenarios either, which makes it seem like its a secondary option to doing your own key rotation.

File: continuous-integration/dev-sec-ops/dependency-container-scanning/dependency_container_scanning.md [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+   - It is hosted on a reputable registry.
+   - The Dockerfile is available, and check for dependencies installed in it.
+   - The image is frequently updated - old images might not contain the latest security updates.
+1. Remove Non-Essential Software - Start with a minimal base image and install only the tools, libraries and configuration files that are required by your application.
+Avoid installing the following tools or remove them if present:
+    - Network tools and clients: e.g., wget, curl, netcat, ssh.
+    - Shells: e.g. sh, bash. Note that removing shells also prevents the use of shell scripts at runtime. Instead, use an executable when possible.
+    - Compilers and debuggers. These should be used only in build and development containers, but never in production containers.
+1. Container images should be immutable - download and include all the required dependencies during the image build.
+
```
Comment (balteravishay) [inline continuous-integration/dev-sec-ops/dependency-container-scanning/dependency_container_scanning.md:30]: seems like the best practices only apply to image scanning and not to software dependency scanning tools.

File: continuous-integration/dev-sec-ops/dependency-container-scanning/dependency_container_scanning.md [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+## Dependency and Container Scanning Frameworks and Tools
+
+Use the tools and pipelines as suggested in [CSEDevSecOps Dependency and Container Scanning Scenarios](https://github.com/microsoft/CSEDevSecOps/tree/master/Scenarios/DependencyContainerScanning).
+
+## Conclusion
+
+A powerful technology such as containers should be used carefully. Install the minimal requirements needed for your application and make sure to maintain it over time by using container and dependencies scanning tools.
```
Comment (balteravishay) [inline continuous-integration/dev-sec-ops/dependency-container-scanning/dependency_container_scanning.md:37]: same comment about software dependency scanning

**Comment by DariuszPorowski** [comment]
Hey @LizaShak Just started review.

File: continuous-integration/dev-sec-ops/secret-management/secrets_rotation.md (hunk 1 above)
Comment (DariuszPorowski) [reply continuous-integration/dev-sec-ops/secret-management/secrets_rotation.md:12]: Agree with @seushermsft

**Review by DariuszPorowski — CHANGES_REQUESTED** [review]
@LizaShak I have only 1 feedback item, aligned with @seushermsft.

File: continuous-integration/dev-sec-ops/secret-management/secrets_rotation.md (hunk 1 above)
Comment (LizaShak) [reply continuous-integration/dev-sec-ops/secret-management/secrets_rotation.md:20]: @seushermsft you are totally right.  Added information about Managed Identities and that they should be our default choice.

**Review by balteravishay — APPROVED** [review]

**Review by seushermsft — APPROVED** [review]

**Review by brockneedscoffee — APPROVED** [review]

File: continuous-integration/dev-sec-ops/dependency-container-scanning/dependency_container_scanning.md [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    - Compilers and debuggers. These should be used only in build and development containers, but never in production containers.
+1. Container images should be immutable - download and include all the required dependencies during the image build.
+1. Scan for vulnerabilities in software dependencies -  today there is likely no software project without some form of external libraries, dependencies or open source.
+While it allows the development team to focus on their application code, the dependency brings forth an expected downside where the security posture of the real application is now resting on it.
+To detect vulnerabilities contained within a project’s dependencies use container scanning tools which as part of their analysis scan the software dependencies (see "Dependency and Container Scanning Frameworks and Tools").
+
+## Dependency and Container Scanning Frameworks and Tools
+
+Use the tools and pipelines as suggested in [CSEDevSecOps Dependency and Container Scanning Scenarios](https://github.com/microsoft/CSEDevSecOps/tree/master/Scenarios/DependencyContainerScanning).
```
Comment (brockneedscoffee) [inline continuous-integration/dev-sec-ops/dependency-container-scanning/dependency_container_scanning.md:36]: Given this is an open source playbook that we share with our customers and those outside Microsoft I don't think we should include this link.  It is not accessible by people who are not part of the Microsoft org, it 404s.

File: continuous-integration/dev-sec-ops/penetration-testing/penetration_testing.md [hunk 5]
```diff
@@ -0,0 +1,22 @@
+# Penetration Testing
+
+A penetration test is a simulated attack against your application to check for exploitable security issues.
+For further reading  see [CSEDevSecOps PenetrationTesting Scenarios](https://github.com/microsoft/CSEDevSecOps/tree/master/Scenarios/PenetrationTesting)
```
Comment (brockneedscoffee) [inline continuous-integration/dev-sec-ops/penetration-testing/penetration_testing.md:4]: Given this is an open source playbook that we share with our customers and those outside Microsoft I don't think we should include this link.  It is not accessible by people who are not part of the Microsoft org, it 404s.

File: continuous-integration/dev-sec-ops/readme.md [hunk 6]
```diff
@@ -0,0 +1,30 @@
+# DevSecOps
+
+## The concept of DevSecOps
+
+DevSecOps or DevOps security is about introducing security earlier in the life cycle of application development (a.k.a shift-left), thus minimizing the impact of vulnerabilities and bringing security closer to development team.
+Further information is available [here](https://github.com/microsoft/CSEPackagedSprints/tree/master/DevSecOps).
```
Comment (brockneedscoffee) [inline continuous-integration/dev-sec-ops/readme.md:6]: Given this is an open source playbook that we share with our customers and those outside Microsoft I don't think we should include this link. It is not accessible by people who are not part of the Microsoft org, it 404s.

File: continuous-integration/dev-sec-ops/readme.md [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+
+## Why
+
+By embracing shift-left mentality, DevSecOps encourages organizations to bridge the gap that often exists between development and security teams to the point where many of the security processes are automated and are effectively handled by the development team.
+
+## CSE Packaged Sprints
+
+DevOps, Identity, and Security have standard foundational requirements across Dev Crew projects. The CSE Packaged sprints are set out to tackle this repetitive work. Adding a packaged sprint offers a pathway to dramatically improve the process of incorporating foundational concepts.
+Further information is available [here](https://github.com/microsoft/CSEPackagedSprints).
```
Comment (brockneedscoffee) [inline continuous-integration/dev-sec-ops/readme.md:15]: 404s unless you are part of the Microsoft org in GitHub

File: continuous-integration/dev-sec-ops/readme.md [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+
+1. [Credential Scanning](./secret-management/credential_scanning.md) - automatically inspecting a project to ensure that no secrets are included in the project's source code.
+1. [Secrets Rotation](./secret-management/secrets_rotation.md) - automated process by which the secret, used by the application, is refreshed and replaced by a new secret.
+1. [Static Code Analysis](./static-code-analysis/static_code_analysis.md) - analyze source code or compiled versions of code to help find security flaws.
+1. [Penetration Testing](./penetration-testing/penetration_testing.md) - a simulated attack against your application to check for exploitable vulnerabilities.
+1. [Container Dependencies Scanning](./dependency-container-scanning/dependency_container_scanning.md) - search for vulnerabilities in container operating systems, language packages and application dependencies.
+
+## Resources
+
+[CSEDevSecOps Scenarios](https://github.com/microsoft/CSEDevSecOps/tree/master/Scenarios)
```
Comment (brockneedscoffee) [inline continuous-integration/dev-sec-ops/readme.md:30]: 404s unless you are part of the Microsoft org in GitHub

File: continuous-integration/dev-sec-ops/secret-management/credential_scanning.md [hunk 9]
```diff
@@ -0,0 +1,38 @@
+# Credential Scanning
+
+Credential scanning is the practice of automatically inspecting a project to ensure that no secrets are included in the project's source code. Secrets include database passwords, storage connection strings, admin logins, service principals, etc.
+For further reading see [CSEDevSecOps Credential Scanning Scenarios](https://github.com/microsoft/CSEDevSecOps/tree/master/Scenarios/CredentialScanning).
```
Comment (brockneedscoffee) [inline continuous-integration/dev-sec-ops/secret-management/credential_scanning.md:4]: 404s unless you are part of the Microsoft org in GitHub

**Review by brockneedscoffee — CHANGES_REQUESTED** [review]
This is really great content.  The Microsoft repo links should probably be removed though since folks who are not part of the Microsoft org cannot access it and this is intended for use outside of Microsoft (e.g. CSE customers, open source community, etc)

**Comment by brockneedscoffee** [comment]
@LizaShak is there an update on this?  Do you need some assistance?

**Comment by LizaShak** [comment]
> @LizaShak is there an update on this? Do you need some assistance?

Hi, we have checked with CSE DevSecOps team to see if those links will be public soon, and they won't in the near two months.So I am working on replacing those links with the relevant information.  Once those links will be public we will create a separate PR for it.

**Comment by fnocera** [comment]
Once #607 is merged in all outstanding PRs including this one will have to make a few changes and merge master to adopt the new MKDocs structure. This main change is that all the md files move into a `docs` folder. Please don't hesitate to reach out with any questions and a number of us are more than happy to help with this process - reach out to @fnocera @TessFerrandez @omri374 and @brockneedscoffee

**Review by brockneedscoffee — APPROVED** [review]

**Pull request merged**